File: lidar/lidar.c
#include <unistd.h>
#include "lidar.h"

int openDevice(char* filename){
  int file;

  if((file = open(filename, O_RDWR)) < 0){
    return -1;
  }

  if((ioctl(file, I2C_SLAVE_FORCE, DEF_ADDR)) < 0){
    close(file);
    return -1;
  }

  /*if(!isOkay(file)){
    close(file);
    return -1;
  }*/
  
  i2c_smbus_write_byte_data(file, CONTROL_REG, RESET_FPGA); 
  usleep(20000);
  return file;
}

int getDeviceStatus(int file){
  return i2c_smbus_read_byte_data(file, STATUS_REG);
}

int isOkay(int file){
  if((getDeviceStatus(file) & DEVICE_OKAY) != DEVICE_OKAY){
    return 0;
  } else {
    return 1;
  }
}

int isReady(int file){

  // device is ready if bit 0 is 0
  if((getDeviceStatus(file) & DEVICE_READY) == 0){
    return 1;
  }

  return 0;
}

int getDistance(int file){
  while(!isReady(file)){} // spin
  
  if(isReady(file)){
    i2c_smbus_write_byte_data(file, CONTROL_REG, AQUISIT_DC);
  } else {
    return -1;
  }

  while(!isReady(file)){} // spin  
  
  return (i2c_smbus_read_byte_data(file, AQUISIT_REG_MSB) << 8) | i2c_smbus_read_byte_data(file, AQUISIT_REG_LSB);
}

File: AutoPilot/lidar.py
from smbus import SMBus
import Adafruit_GPIO.FT232H as FT232H
import time, threading

class Lidar(threading.Thread):
    __DEF_ADDR = 0x62

    # register addresses
    __CONTROL_REG = 0x00
    __STATUS_REG = 0x01
    __AQUISIT_REG_MSB = 0x0f
    __AQUISIT_REG_LSB = 0x10
    __SERIAL_REG = 0x96

    # commands
    __RESET_FPGA = 0x00
    __AQUISIT_DC = 0x04
    __AQUISIT_NODC = 0x03

    # status register bits
    __MEASURE_ERROR = 0x40
    __DEVICE_OKAY = 0x20
    __INVALID_SIGNAL = 0x08
    __DEVICE_READY = 0x01

    distance = -1

    def __init__(self):
        super(Lidar, self).__init__()
        self.lock = threading.RLock()
        FT232H.use_FT232H()
        ft232h = FT232H.FT232H()
        self.bus = FT232H.I2CDevice(ft232h, self.__DEF_ADDR)
        if self.bus.ping():
            print "Connected"
        self.bus.write8(self.__CONTROL_REG, self.__RESET_FPGA)
        time.sleep(0.02)
        self._stop = threading.Event()
        self.start()

    def getDeviceStatus(self):
        return self.bus.readU8(self.__STATUS_REG)

    def isOkay(self):
        if (self.getDeviceStatus() & self.__DEVICE_OKAY) != self.__DEVICE_OKAY:
            return False
        else:
            return True

    def isReady(self):
        if (self.getDeviceStatus() & self.__DEVICE_READY) == 0:
            return True
        else:
            return False

    def checkMeasureError(self):
        if (self.getDeviceStatus() & self.__MEASURE_ERROR):
            return True
        else:
            return False

    def readDistance(self):
        return self.distance

    def run(self):
        self._stop.clear()
        while self._stop.isSet() is False:
            self.lock.acquire()
            self.distance = self.__getDistance()
            self.lock.release()
            time.sleep(0.02)

    def stop(self):
        self._stop.set()

    def acquireLock(self):
        self.lock.acquire()

    def releaseLock(self):
        self.lock.release()

    def __getDistance(self):
        self.__spin_while_not_ready()
        self.bus.write8(self.__CONTROL_REG, self.__AQUISIT_DC)
        self.__spin_while_not_ready()
        return (self.bus.readU8(self.__AQUISIT_REG_MSB) << 8) | (self.bus.readU8(self.__AQUISIT_REG_LSB))

    def __spin_while_not_ready(self):
        while (not self.isReady()):
            continue